pkg/scheduler/plugins/nodegroup/nodegroup.go

```diff
@@ -0,0 +1,261 @@
+/*
+Copyright 2023 The Volcano Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodegroup
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	batch "volcano.sh/apis/pkg/apis/batch/v1alpha1"
+	sch "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
+	"volcano.sh/volcano/pkg/scheduler/api"
+	"volcano.sh/volcano/pkg/scheduler/framework"
+)
+
+const (
+	// PluginName indicates name of volcano scheduler plugin.
+	PluginName       = "nodegroup"
+	NodeGroupNameKey = "volcano.sh/nodegroup-name"
+)
+
+type nodeGroupPlugin struct {
+	// Arguments given for the plugin
+	pluginArguments framework.Arguments
+}
+
+// New function returns prioritize plugin object.
+func New(arguments framework.Arguments) framework.Plugin {
+	return &nodeGroupPlugin{pluginArguments: arguments}
+}
+
+func (pp *nodeGroupPlugin) Name() string {
+	return PluginName
+}
+
+type queueGroupAffinity struct {
+	queueGroupAntiAffinityRequired  map[string][]string
+	queueGroupAntiAffinityPreferred map[string][]string
+	queueGroupAffinityRequired      map[string][]string
+	queueGroupAffinityPreferred     map[string][]string
+}
+
+func NewQueueGroupAffinity() queueGroupAffinity {
+	return queueGroupAffinity{
+		queueGroupAntiAffinityRequired:  make(map[string][]string, 0),
+		queueGroupAntiAffinityPreferred: make(map[string][]string, 0),
+		queueGroupAffinityRequired:      make(map[string][]string, 0),
+		queueGroupAffinityPreferred:     make(map[string][]string, 0),
+	}
+}
+func (q queueGroupAffinity) predicate(queue, group string) error {
+	if len(queue) == 0 {
+		return nil
+	}
+	flag := false
+	if q.queueGroupAffinityRequired != nil {
+		if groups, ok := q.queueGroupAffinityRequired[queue]; ok {
+			if contains(groups, group) {
+				flag = flag || true
```

Comment:
`||` is necessary here, and `preferred` items should not be considered in predict.